crates/ra_hir/src/ty/traits.rs

```diff
@@ -1,115 +1,428 @@
-//! Stuff that will probably mostly replaced by Chalk.
-use std::collections::HashMap;
+//! Chalk integration.
+use std::sync::{Arc, Mutex};
 
-use crate::{db::HirDatabase, generics::HasGenericParams};
-use super::{TraitRef, Substs, infer::{TypeVarId, InferTy}, Ty};
+use chalk_ir::{TypeId, TraitId, StructId, ImplId, TypeKindId, ProjectionTy, Parameter, Identifier, cast::Cast};
+use chalk_rust_ir::{AssociatedTyDatum, TraitDatum, StructDatum, ImplDatum};
 
-// Copied (and simplified) from Chalk
+use crate::{Crate, Trait, db::HirDatabase, HasGenericParams, ImplBlock};
+use super::{TraitRef, Ty, ApplicationTy, TypeCtor, Substs, infer::Canonical};
 
-#[derive(Clone, Debug, PartialEq, Eq)]
-/// A (possible) solution for a proposed goal. Usually packaged in a `Result`,
-/// where `Err` represents definite *failure* to prove a goal.
-pub enum Solution {
-    /// The goal indeed holds, and there is a unique value for all existential
-    /// variables.
-    Unique(Substs),
+#[derive(Debug, Copy, Clone)]
+struct ChalkContext<'a, DB> {
+    db: &'a DB,
+    krate: Crate,
+}
 
-    /// The goal may be provable in multiple ways, but regardless we may have some guidance
-    /// for type inference.
-    Ambig(Guidance),
+pub(crate) trait ToChalk {
+    type Chalk;
+    fn to_chalk(self, db: &impl HirDatabase) -> Self::Chalk;
+    fn from_chalk(db: &impl HirDatabase, chalk: Self::Chalk) -> Self;
```

Comment:
the strage thing about these signatures is that `db`s could be potentially of different types. Perhaps something like `ToChalk<DB>` would be more correct, but also more complex. So I think the current signatures are ok.

In general, being generic over `DB` causes quite a bit of boilerplate everywhere. I wonder if we should byte the bullet and switch to `&dyn DB` by default everywhere?

Reply:
Both methods need to be able to deal with any database implementation; changing this to move the type parameter to the trait doesn't really make sense to me since it'd make it possible to implement the trait just for a specific database implementation.

Reply:
And I think boilerplate-wise at least in this case `&dyn DB` wouldn't do anything, although it might reduce compile time...